Added Edit Message Functionality

earthtone opened the pull request:
Refactored "remove" functionality to take list index, instead of
computing list index based on text matching. This allows less callback
binding and relying on the "this" keyword.

"Edit" functionality is essentially the same as the "remove"
functionality, but additionally assigning the given messages value to
tempTxt to be rendered in the MessageInput & DisplayText components.

Refactored directory structure to differentiate between "dumb"
components that are only concerned with rendering out data pumped into
them, and "smart" containers that evaluate/process data and state.

Changes to be committed:

-	modified:   app/javascript/bundles/App/actions/index.js
-	new file:   app/javascript/bundles/App/components/MessageList.js
-	new file:   app/javascript/bundles/App/components/TextDisplay.js
-	deleted:    app/javascript/bundles/App/containers/EditTextDisplay.js
-	new file:   app/javascript/bundles/App/containers/MessageInput.js
-	deleted:    app/javascript/bundles/App/containers/MessageList.js
-	new file:   app/javascript/bundles/App/containers/Messages.js
-	modified:   app/javascript/bundles/App/containers/SomethingSmall.js
-	deleted:    app/javascript/bundles/App/containers/TextInputControl.js
-	modified:   app/javascript/bundles/App/reducers/basic.js
-	modified:   app/javascript/bundles/App/startup/BasicFlow.js

Commit:
Added Edit Message Functionality

Refactored "remove" functionality to take list index, instead of
computing list index based on text matching. This allows less callback
binding and relying on the "this" keyword.

"Edit" functionality is essentially the same as the "remove"
functionality, but additionally assigning the given messages value to
tempTxt to be rendered in the MessageInput & DisplayText components.

Refactored directory structure to differentiate between "dumb"
components that are only concerned with rendering out data pumped into
them, and "smart" containers that evaluate/process data and state.

Changes to be committed:

-	modified:   app/javascript/bundles/App/actions/index.js
-	new file:   app/javascript/bundles/App/components/MessageList.js
-	new file:   app/javascript/bundles/App/components/TextDisplay.js
-	deleted:    app/javascript/bundles/App/containers/EditTextDisplay.js
-	new file:   app/javascript/bundles/App/containers/MessageInput.js
-	deleted:    app/javascript/bundles/App/containers/MessageList.js
-	new file:   app/javascript/bundles/App/containers/Messages.js
-	modified:   app/javascript/bundles/App/containers/SomethingSmall.js
-	deleted:    app/javascript/bundles/App/containers/TextInputControl.js
-	modified:   app/javascript/bundles/App/reducers/basic.js
-	modified:   app/javascript/bundles/App/startup/BasicFlow.js

emorris commented on app/javascript/bundles/App/components/MessageList.js:
```diff
@@ -0,0 +1,37 @@
+      <h1>Messages Added</h1>
+      <ol>
+        {
+          messages.map((m, i) => {
```
as a design... i lean toward having  a component that would be message that takes in the message data... and leave the construction of the list of objects at the same level you are getting the data out of state.. this avoids having to pass around large chunks data..  also one would argue that the way you wrote this that messagelist should be a container since it wants what is already in global state

earthtone replied:
Yeah, I'm inclined to agree with you on having a "dumb" component that is responsible for rendering __a message__ and then having a container handle all the mapping.

emorris commented on app/javascript/bundles/App/containers/MessageInput.js:
```diff
@@ -0,0 +1,50 @@
+ *
+ * */
+
+function MessageInput ({ messageText, postMessage, setMessageText  }) {
```
not sure about this .. you are not extending component so you might be loosing useful things by writing the code this way. also you boxing yourself in with this kind of syntax since it really doesn't allow greater expansion of these components... i avoid using this because, once you start having more features you have to rewrite all of this

earthtone replied:
Yeah, this is a point of confusion for me in regards to React. I don't quite understand with any real depth how to take advantage of lifecycle hooks. Especially when you add Redux to the mix. Maybe that'll be more clear when we start integrating more of the Rails/API stuff.

Personally, I try to avoid relying in the "this" keyword as much as possible because it's super confusing for most people. Using class syntax encourages that, and when you're three layers removed from your action creator/event handler and you've had to bind "this" twice, I don't really see the point. Reverting the code to a function literal is just my way of making things more explicit and deterring the temptation to rely on "this", but also making the code reasonable if you _have_ to use "this" since because of how it's scoped.

I see your point about boxing myself in, though. I think I'm used to writing code that older/unfamiliar developers have to read through, so I may take it a bit too far sometimes.

Commit:
Updated Syntax for Ehancement

As per feedback, revert "smart" components to Class syntax, as future
enahancements may require component methods or lifecycle hooks. May be a
more flexible solution towards smart/dumb architectures.

Created Message component so that it can be "dumber", and allow
MessageList component to worry about iteration/mapping.

Changes to be committed:

-	new file:   app/javascript/bundles/App/components/Message.js
-	deleted:    app/javascript/bundles/App/components/MessageList.js
-	modified:   app/javascript/bundles/App/components/TextDisplay.js
-	modified:   app/javascript/bundles/App/containers/MessageInput.js
-	new file:   app/javascript/bundles/App/containers/MessageList.js
-	deleted:    app/javascript/bundles/App/containers/Messages.js
-	modified:   app/javascript/bundles/App/containers/SomethingSmall.js
